internal/syncqueues/priority_queue_with_stats.go

```diff
@@ -0,0 +1,148 @@
+package syncqueues
+
+import (
+	"fmt"
+	"log/slog"
+	"sync"
+	"time"
+)
+
+// Modified priority queue implementation based on heap
+//  - priority based on minimal value of ModifiedAt
+//  - semantics of insert operation is redefined
+// 		we insert element with non-existing subaccountID,
+// 		we update element with existing subaccountID if it is outdated
+
+type SubaccountAwarePriorityQueueWithCallbacks struct {
+	elements     []ElementWrapper
+	idx          map[string]int
+	size         int
+	mutex        sync.Mutex
+	log          *slog.Logger
+	eventHandler *EventHandler
+}
+
+func NewPriorityQueueWithCallbacks(log *slog.Logger, eventHandler *EventHandler) *SubaccountAwarePriorityQueueWithCallbacks {
+	return NewPriorityQueueWithCallbacksForSize(log, eventHandler, DefaultQueueSize)
+}
+
+func NewPriorityQueueWithCallbacksForSize(log *slog.Logger, eventHandler *EventHandler, initialQueueSize int) *SubaccountAwarePriorityQueueWithCallbacks {
+	return &SubaccountAwarePriorityQueueWithCallbacks{
+		elements:     make([]ElementWrapper, initialQueueSize),
+		idx:          make(map[string]int),
+		size:         0,
+		log:          log,
+		eventHandler: eventHandler,
+	}
+}
+
+func (q *SubaccountAwarePriorityQueueWithCallbacks) Insert(element QueueElement) {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	e := ElementWrapper{
+		QueueElement: element,
+		entryTime:    time.Now().UnixNano(),
+	}
+
+	if q.size == cap(q.elements) {
+		newElements := make([]ElementWrapper, q.size*2)
+		copy(newElements, q.elements)
+		q.elements = newElements
+		q.log.Info(fmt.Sprintf("Queue is full, resized to %v", q.size*2))
+	}
+	if idx, ok := q.idx[e.SubaccountID]; ok {
+		if q.elements[idx].ModifiedAt > e.ModifiedAt {
+			// event is outdated, do not insert
+			return
+		}
+		// extraction and re-insertion
+		q.elements[idx].ModifiedAt = -1
+		q.siftUpFrom(idx)
+		// updated element is at the top of the heap
+		// remove it
+		q.elements[0] = q.elements[q.size-1]
+		q.size--
+		q.siftDown()
+	} else {
+		q.idx[e.SubaccountID] = q.size
+	}
+	q.elements[q.size] = e
+	q.size++
+	q.siftUp()
+
+	if q.eventHandler != nil && q.eventHandler.OnInsert != nil {
+		q.eventHandler.OnInsert(q.size)
+	}
+}
+
+func (q *SubaccountAwarePriorityQueueWithCallbacks) Extract() QueueElement {
```

Comment:
maybe Extract should return a bool which indicates if there was an item or not

Reply:
Changes pushed